xarray/testing/strategies.py

```diff
@@ -0,0 +1,343 @@
+from collections.abc import Hashable, Mapping, Sequence
+from typing import Any, Protocol, Union
+
+import hypothesis.extra.numpy as npst
+import hypothesis.strategies as st
+import numpy as np
+
+import xarray as xr
+from xarray.core.types import T_DuckArray
+
+__all__ = [
+    "numeric_dtypes",
+    "names",
+    "dimension_names",
+    "dimension_sizes",
+    "attrs",
+    "variables",
+]
+
+
+class ArrayStrategyFn(Protocol):
+    def __call__(
+        self, *, shape: tuple[int, ...] = None, dtype: np.dtype = None, **kwargs
+    ) -> st.SearchStrategy[T_DuckArray]:
+        ...
+
+
+# required to exclude weirder dtypes e.g. unicode, byte_string, array, or nested dtypes.
+def numeric_dtypes() -> st.SearchStrategy[np.dtype]:
+    """
+    Generates only those numpy dtypes which xarray can handle.
+
+    Requires the hypothesis package to be installed.
+    """
+
+    return (
+        npst.integer_dtypes()
+        | npst.unsigned_integer_dtypes()
+        | npst.floating_dtypes()
+        | npst.complex_number_dtypes()
+    )
```

Comment:
It might well be the case that xarray is robust enough to handle all of these, in which case we don't need this strategy. I would suggest we merge is for now, test some NamedArray functionality with it, and if that looks good then consider relaxing this.